# **2.3.4 - Bài tập thực hành 2: Xây dựng mô hình Bayes ngây thơ trên tập dữ liệu mushroom**

## **1. Tải và khám phá dữ liệu**

In [1]:
import pandas as pd

# Đọc dữ liệu
data = pd.read_csv('dataset/mushrooms.csv')  # thay bằng đường dẫn thực tế nếu khác

# Xem thông tin tổng quan
print(data.info())
print(data.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

## **2. Tiền xử lý dữ liệu**

In [2]:
# Kiểm tra giá trị thiếu
print(data.isnull().sum())

# Nếu có dấu '?' đại diện cho giá trị thiếu, thay thế bằng NaN
data.replace('?', pd.NA, inplace=True)

# Xóa các dòng có giá trị thiếu (hoặc xử lý khác nếu muốn)
data.dropna(inplace=True)

# Mã hóa biến phân loại (toàn bộ cột đều là dạng chữ)
data_encoded = pd.get_dummies(data, drop_first=True)


class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64


## **3. Tách dữ liệu và chia tập**

In [3]:
from sklearn.model_selection import train_test_split

# Cột mục tiêu là 'class' (e: edible, p: poisonous)
X = data_encoded.drop('class_p', axis=1)  # giữ lại 'class_p' làm nhãn
y = data_encoded['class_p']  # 1: poisonous, 0: edible

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


## **4. Huấn luyện mô hình Naïve Bayes**

In [4]:
from sklearn.naive_bayes import GaussianNB

nb_model = GaussianNB()
nb_model.fit(X_train, y_train)


GaussianNB()

## **5. Dự đoán và đánh giá mô hình**

In [5]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

y_pred = nb_model.predict(X_test)

print("Naïve Bayes Accuracy:", accuracy_score(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))


Naïve Bayes Accuracy: 0.995571302037201
[[700   5]
 [  0 424]]
              precision    recall  f1-score   support

       False       1.00      0.99      1.00       705
        True       0.99      1.00      0.99       424

    accuracy                           1.00      1129
   macro avg       0.99      1.00      1.00      1129
weighted avg       1.00      1.00      1.00      1129



**Nhận xét:** Mô hình Naïve Bayes đạt độ chính xác gần như tuyệt đối (99.56%), phân loại rất tốt cả nấm độc và nấm ăn được. Precision và recall đều xấp xỉ 1.00, cho thấy mô hình hoạt động hiệu quả và đáng tin cậy trên tập dữ liệu này.

## **6. So sánh với các mô hình khác**

In [9]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report

# Khởi tạo các mô hình
log_model = LogisticRegression(max_iter=1000)
tree_model = DecisionTreeClassifier()
knn_model = KNeighborsClassifier()

# Huấn luyện
log_model.fit(X_train, y_train)
tree_model.fit(X_train, y_train)
knn_model.fit(X_train, y_train)

# Dự đoán
log_pred = log_model.predict(X_test)
tree_pred = tree_model.predict(X_test)
knn_pred = knn_model.predict(X_test)

# Đánh giá
print("+ Logistic Regression Accuracy:", accuracy_score(y_test, log_pred))
print(classification_report(y_test, log_pred))

print("+ Decision Tree Accuracy:", accuracy_score(y_test, tree_pred))
print(classification_report(y_test, tree_pred))

print("+ KNN Accuracy:", accuracy_score(y_test, knn_pred))
print(classification_report(y_test, knn_pred))


+ Logistic Regression Accuracy: 1.0
              precision    recall  f1-score   support

       False       1.00      1.00      1.00       705
        True       1.00      1.00      1.00       424

    accuracy                           1.00      1129
   macro avg       1.00      1.00      1.00      1129
weighted avg       1.00      1.00      1.00      1129

+ Decision Tree Accuracy: 1.0
              precision    recall  f1-score   support

       False       1.00      1.00      1.00       705
        True       1.00      1.00      1.00       424

    accuracy                           1.00      1129
   macro avg       1.00      1.00      1.00      1129
weighted avg       1.00      1.00      1.00      1129

+ KNN Accuracy: 1.0
              precision    recall  f1-score   support

       False       1.00      1.00      1.00       705
        True       1.00      1.00      1.00       424

    accuracy                           1.00      1129
   macro avg       1.00      1.00      1.0

**Nhận xét:** Tất cả các mô hình — Logistic Regression, Decision Tree, KNN — đều đạt độ chính xác tuyệt đối (100%), nhỉnh hơn Naïve Bayes (99.56%). Tuy Naïve Bayes vẫn rất hiệu quả, nhưng các mô hình còn lại cho kết quả phân loại hoàn hảo hơn trên tập dữ liệu nấm.